### PR TITLE
Add backend state summary to Triton logs

### PR DESCRIPTION
**Pull request by Tabrizian**
A screenshot of how the logs will look like:

image

### CODE REVIEW

**Comment by CoderHam** [comment]
If you add `server_id`, `server_version` and `server_extension` to the table it can be removed from server.cc

File: src/core/server.cc [hunk 1]
```diff
@@ -109,6 +113,90 @@ InferenceServer::InferenceServer()
...
+  auto backend_map = triton_backend_manager.BackendMap();
+
+  for (const auto& backend_pair : backend_map) {
+    std::vector<std::string> backend_record;
+    auto backend = backend_pair.second.lock();
+    if (backend != nullptr) {
+      backend_record.emplace_back(backend->Name());
+      const char* config_message;
+      size_t config_message_size;
+      backend->BackendConfig().Serialize(&config_message, &config_message_size);
```
Comment (CoderHam) [inline src/core/server.cc:185]: Do we want to print out the entire config message? I think this is necessary only if the model config is autofilled.

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: src/core/server.cc [hunk 2]
```diff
@@ -47,6 +47,10 @@
 #include "src/core/pinned_memory_manager.h"
 #include "src/core/server.h"
 
+#include "src/backends/backend/triton_backend_manager.h"
+
```
Comment (CoderHam) [inline src/core/server.cc:51]: Remove extra lines after and before triton_backend_manager.h . 
Also if you check 
`#include "src/core/server.h"`
Is mistakenly added twice. Remove second occurrence.

**Comment by deadeyegoodwin** [comment]
@Tabrizian In backend table "path" and "config" are swapped. 
Can you put the backend table before the model table.
Put version and extensions in the first table and then you can remove them from the other logging.

File: src/backends/backend/triton_backend_manager.cc [hunk 3]
```diff
@@ -327,15 +327,29 @@ TRITONBACKEND_BackendSetState(TRITONBACKEND_Backend* backend, void* state)
 //
 // TritonBackendManager
 //
+
+TritonBackendManager&
+TritonBackendManager::Instance()
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.cc:332]: Instead of breaking the encapsulation of the class by exposing these internals, just add a BackendState() method that fills a map from backend-name to a record of state information (path and config).

File: src/core/server.cc [hunk 4]
```diff
@@ -109,6 +113,90 @@ InferenceServer::InferenceServer()
...
+            ModelReadyStateString(model_status_pair.first);
+
+        if (model_status_pair.second != "") {
+          model_status += ": " + model_status_pair.second;
+        }
+
+        model_record.emplace_back(model_name);
+        model_record.emplace_back(model_version);
+        model_record.emplace_back(model_status);
+        models_table->insert_row(model_record);
```
Comment (CoderHam) [inline src/core/server.cc:155]: You are re-using the same model_record for all versions in the model_version_map. Shift declaration of model record inside for loop.

File: src/core/server.cc [hunk 5]
```diff
@@ -109,6 +113,90 @@ InferenceServer::InferenceServer()
...
+  triton::common::TablePrinter::Create(&models_table, model_headers);
+
+  for (const auto& model_state : model_states) {
+    std::vector<std::string> model_record;
+    auto model_version_map = model_state.second;
+    std::string model_name = model_state.first;
+
+    // If model_version_map size is zero, no version is found for this model
+    if (model_version_map.size() == 0) {
+      model_record.emplace_back(model_name);
```
Comment (CoderHam) [inline src/core/server.cc:137]: Shift declaration of model_record to here

File: src/backends/backend/triton_backend_manager.h [hunk 6]
```diff
@@ -143,8 +143,24 @@ class TritonBackendManager {
...
+  static TritonBackendManager& Instance();
+  static std::mutex& Mutex()
+  {
+    static std::mutex m;
+    return m;
+  }
+
+  TritonBackendManager() {}
+
+  TritonBackendManager(TritonBackendManager const&) = delete;
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.h:155]: Use DISALLOW_COPY_AND_ASSIGN macro from constants.h.
Put macro at the start of the "private:" section.

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: src/core/server.cc [hunk 7]
```diff
@@ -189,12 +277,15 @@ InferenceServer::Init()
       // failure is due to a model not loading correctly so we just
       // continue if not exiting on error.
       ready_state_ = ServerReadyState::SERVER_READY;
+      PrintSummary();
     }
-    return status;
+  } else {
+    ready_state_ = ServerReadyState::SERVER_READY;
+    status = Status::Success;
```
Comment (deadeyegoodwin) [inline src/core/server.cc:284]: Setting status here is redundant

File: src/core/server.h [hunk 8]
```diff
@@ -97,6 +97,9 @@ class InferenceServer {
   Status ModelReadyVersions(
       std::map<std::string, std::vector<int64_t>>* model_versions);
 
+  // Print backends and models summary
+  void PrintSummary();
```
Comment (deadeyegoodwin) [inline src/core/server.h:101]: PrintBackendAndModelSummary()

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: src/core/server.cc (hunk 1 above)
Comment (Tabrizian) [reply src/core/server.cc:185]: Usually, backends do not have a lot of config flags. I can add an if statement to print it only if it is being auto-filled. 

/cc @deadeyegoodwin

Comment (deadeyegoodwin) [reply src/core/server.cc:185]: I think @CoderHam  was confusing the backend config with model config. I think what you have it ok.

**Comment by deadeyegoodwin** [comment]
@Tabrizian  There is also a cuda pool you should include in the options table

File: src/core/server.cc (hunk 1 above)
Comment (CoderHam) [reply src/core/server.cc:185]: That's right I was mistaken. Marking this as resolved.

**Comment by Tabrizian** [comment]
@deadeyegoodwin CUDA pool will be shown if there is anything inside the map. Currently, the map is empty.

File: src/backends/backend/triton_backend_manager.h (hunk 6 above)
Comment (Tabrizian) [reply src/backends/backend/triton_backend_manager.h]: I removed exposing the Singleton instance altogether. Marking this as resolved.

**Comment by Tabrizian** [comment]
My bad, it seems like I was using the wrong flag. I fixed the CUDA memory pool.

**Comment by Tabrizian** [comment]
image
 Above, is the screenshot of the latest version of the backend. If the screen size is small, the table will look like below:

image

File: src/backends/backend/triton_backend_manager.cc [hunk 9]
```diff
@@ -356,7 +360,19 @@ TritonBackendManager::CreateBackend(
       name, dir, libpath, backend_cmdline_config, backend));
   singleton_manager.backend_map_.insert({libpath, *backend});
 
+  const char* backend_config;
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.cc:363]: What about when a backend unloads? Don't try to update the BackedState incrementally, just calculate it each time BackendState() is called.

File: src/backends/backend/triton_backend_manager.h [hunk 10]
```diff
@@ -143,9 +143,14 @@ class TritonBackendManager {
       const BackendCmdlineConfig& backend_cmdline_config,
       std::shared_ptr<TritonBackend>* backend);
 
+  static const std::unordered_map<std::string, std::vector<std::string>>&
+  BackendState();
+
  private:
-  std::mutex mu_;
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.h:147]: Why did you change this mutex name? As a convention we typically use mu_. Can you put it back like it was?

File: src/backends/backend/triton_backend_manager.h [hunk 11]
```diff
@@ -143,9 +143,14 @@ class TritonBackendManager {
...
+  static std::unordered_map<std::string, std::vector<std::string>>
+      backend_state_;
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.h:151]: Don't need this.

File: src/backends/backend/triton_backend_manager.cc [hunk 12]
```diff
@@ -356,7 +360,19 @@ TritonBackendManager::CreateBackend(
...
+  backend_state_.insert(
+      {name, std::vector<std::string>{libpath, backend_config}});
+
   return Status::Success;
 }
 
+const std::unordered_map<std::string, std::vector<std::string>>&
+TritonBackendManager::BackendState()
+{
+  return TritonBackendManager::backend_state_;
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.cc:375]: Create backend-state here using backend_map_. Be sure to lock mutex.

File: src/core/model_repository_manager.cc [hunk 13]
```diff
@@ -699,7 +699,7 @@ ModelRepositoryManager::BackendLifeCycle::Load(
 
   switch (backend_info->state_) {
     case ModelReadyState::READY:
-      LOG_INFO << "re-loading: " << model_name << ":" << version;
+      LOG_VERBOSE(1) << "re-loading: " << model_name << ":" << version;
```
Comment (deadeyegoodwin) [inline src/core/model_repository_manager.cc:702]: Revert these logging changes

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

**Comment by Tabrizian** [comment]
@deadeyegoodwin I had to change TritonBackendManager singleton a little bit to make the instance available to the internal static methods (e.g. `BackendState`). I changed the visibility of all those methods to ensure that it doesn't break the encapsulations.

File: src/backends/backend/triton_backend_manager.cc [hunk 14]
```diff
@@ -327,14 +327,20 @@ TRITONBACKEND_BackendSetState(TRITONBACKEND_Backend* backend, void* state)
 //
 // TritonBackendManager
 //
+TritonBackendManager&
+TritonBackendManager::Instance_()
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.cc:331]: Singleton(). Putting _ at the end of a function name is not conforming to our coding standards.

File: src/backends/backend/triton_backend_manager.cc [hunk 15]
```diff
@@ -359,4 +365,31 @@ TritonBackendManager::CreateBackend(
   return Status::Success;
 }
 
+std::unique_ptr<std::unordered_map<std::string, std::vector<std::string>>>
+TritonBackendManager::BackendState()
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.cc:369]: Convention would be to return Status from function and add std::unique_ptr<...>* as argument which you then initialize with the map.

File: src/backends/backend/triton_backend_manager.h [hunk 16]
```diff
@@ -143,9 +144,19 @@ class TritonBackendManager {
       const BackendCmdlineConfig& backend_cmdline_config,
       std::shared_ptr<TritonBackend>* backend);
 
+  static std::unique_ptr<
+      std::unordered_map<std::string, std::vector<std::string>>>
+  BackendState();
+
+  TritonBackendManager(){};
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.h:151]: Why is this needed as it is empty? It it is needed add it in the private section as:
TritonBackendManager() = default

File: src/backends/backend/triton_backend_manager.h [hunk 17]
```diff
@@ -143,9 +144,19 @@ class TritonBackendManager {
...
+
  private:
+  static std::unordered_map<std::string, std::vector<std::string>>
+      backend_state_;
+  static TritonBackendManager& Instance_();
   std::mutex mu_;
   std::unordered_map<std::string, std::weak_ptr<TritonBackend>> backend_map_;
+  DISALLOW_COPY_AND_ASSIGN(TritonBackendManager);
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.h:159]: Put this right after "private:"

File: src/backends/backend/triton_backend_manager.h [hunk 18]
```diff
@@ -143,9 +144,19 @@ class TritonBackendManager {
...
       std::shared_ptr<TritonBackend>* backend);
 
+  static std::unique_ptr<
+      std::unordered_map<std::string, std::vector<std::string>>>
+  BackendState();
+
+  TritonBackendManager(){};
+
  private:
+  static std::unordered_map<std::string, std::vector<std::string>>
```
Comment (deadeyegoodwin) [inline src/backends/backend/triton_backend_manager.h:154]: Should not be needed. Is it even used?

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: src/backends/backend/triton_backend_manager.h (hunk 16 above)
Comment (Tabrizian) [reply src/backends/backend/triton_backend_manager.h:152]: If I don't add this, the deleted copy constructor will be picked up as the default constructor instead of this constructor.

**Review by deadeyegoodwin — APPROVED** [review]
